### PR TITLE
Add export & import for repeating instruments and events

### PR DESCRIPTION
**Pull request by JuliaSprenger**
With the new test framework I am not sure how / where to integrate a dedicated test for these functions. Any advise on this?

### CODE REVIEW

**Comment by JuliaSprenger** [comment]
I don't think the failing checks are related to my changes, but rather to some server token permissions not granted for tests originating from a fork. @pwildenhain Any way I can fix this or is this an issue of the GH Actions configuration?

**Comment by pwildenhain** [comment]
You are correct, it's the fact that your account can't access repo secrets 

it's happening with the doctests, but what's confusing to me is that I thought I had configured these to skip on forks though, I must've messed up the logic 

If you check out .github/workflows/ci.yml specifically at line 40 you'll see the skip logic I'm talking about

Looks like I'll have to tweak that to move this forward.

Does everything work when you run pytest locally?

**Comment by JuliaSprenger** [comment]
Locally using pytest I get some pylint issues, not sure if these will also pop up in the CI. I addition the `test_empty_json_is_still_a_problem_for_other_methods` test also fails, but I am not sure how that relates to my changes...

Regarding the conditional run of the doctests, maybe you have to use a different syntax (`${{ github.repository == 'redcap-tools/PyCap' }}`) like shown [here](https://docs.github.com/en/actions/learn-github-actions/environment-variables#using-contexts-to-access-environment-variable-values).

**Comment by pwildenhain** [comment]
Yea looks like we might need to use github.ref or github.actor

As a test, can you change that line to only run if github.actor == 'pwildenhain' ?

That's a little hard coded for my usual taste but at least we can get this PR to pass. Down the road I can change this to only run if the ref starts with redcap-tools or something like that (aka don't run if PR originated from a fork)

**Comment by pwildenhain** [comment]
Hmmmm it's appropriately skipping the integration tests, but it looks like it's still trying to build the doctest fixture and that's where it's freaking out.

Really appreciate all your contributions, I'll hopefully be able to provide a lot of fixes once I'm in the office today/and probably bleeding into tomorrow as well

**Comment by pwildenhain** [comment]
Tested locally and everything worked just fine. I'll cut a new pre-release based off this which you can test off to ensure this error is resolved @JuliaSprenger

**Comment by pwildenhain** [comment]
See JuliaSprenger/PyCap#1 for some proposed changes to this PR that I think will allow the CI to run 🤞🏻

**Comment by pwildenhain** [comment]
> I thought I fixed the `typing-extensions` dependencies already via #212, not sure why it is still missing in the CI environment...

I ended up getting rid of `typing-extensions` all together when dropping support for python 3.7 since the types I was using it for entered the stdlib in python 3.8


🎉 So awesome that the CI passes now 🙌🏻 

Would you like to try your hand at adding some unit tests? If so I can walk you through the architecture -- if not I can add them myself.

**Comment by JuliaSprenger** [comment]
The tests fail with a `KeyError: 'repeatingFormsEvents'`. Maybe there's more changes to the long_project.xml needed to be able to export repeating forms? @pwildenhain Any ideas?

**Comment by JuliaSprenger** [comment]
Hi @pwildenhain I copied some of the repeatingFormsEvents tests to be unit tests and added the corresponding lines for the export in the callback_utils. However, I am not sure how these utils are distinguishing between export and import. For this reason the `test_import_export_repeating_forms` still fails. Any hint on how to add a valid response also for the import?

**Comment by pwildenhain** [comment]
You're so close here, that I am just gonna close the gap (see https://github.com/JuliaSprenger/PyCap/pull/3) and I'll add a more in-depth explanation below

> However, I am not sure how these utils are distinguishing between export and import.

This points to a problem with the REDCap API -- **everything** is a `POST` request. Normally exports would fall under `GET` while imports would fall under `POST` and it would be a lot easier to distinguish.

How REDCap distinguishes instead is the presence of the `data` parameter in the payload body, since that is only a valid parameter for imports. So we mimic that in our callback. Example linked below: https://github.com/redcap-tools/PyCap/blob/8a7d169b92b9405365fe3ccf766bb4fb03dca9fb/tests/unit/callback_utils.py#L135-L153

Basically we just check the for the `"data"` key in the `data` portion of the request body (a little confusing 😵 ). And we check for that _first_ so that way if we don't find it, we know we're dealing with an export.

It's kinda neat, gives a little bit of insight into how a real REDCap server needs to handle these types of requests -- but definitely not intuitive 😂

**Comment by JuliaSprenger** [comment]
Thanks for merging and all of the advise you gave in this PR. I learned quite a bit :)
